---
title: "Using AWK and R to parse 25TB"
author: Nick Strayer
date: '2019-06-04'
slug: using_awk_and_R_to_parse_25tb
categories: ["big data", "awk", "data cleaning"]
tags:  ["big data", "awk", "data cleaning"]
excerpt: "Recently I was tasked with parsing 25tb of raw genotype data. This is the story of how I brought the query time and cost down from 8 mins and $20 to 0.1sec and less than a penny."
---


<style>
.lesson_learned {
  color: #238443;
}
.lesson_learned strong {
  color: #005a32;
}

.photo_caption {
  font-size: 0.8em;
}
</style>

__How to read this post:__ I sincerely appologize for how long and rambling the following text is. To speed up skimming of it for those who have better things to do with their time I have started most sections with a ["Lesson learned"]{.lesson_learned} blurb that boils down the takeaway from the following text into a sentence or two. 

__Just show me the sollution!__ If you just want to see how I ended up solving the task jump to the section "_Getting Creative_", but I honestly think the failures are more interesting/valuable. 

## Intro
Recently I was put in charge of setting up a workflow for dealing with a large amount of raw DNA sequencing (well technically a SNP chip) data for my lab.  The goal was to be able to quickly get data for a given genetic location (called a SNP) for use for modeling etc. I managed to get a workflow setup involving vanilla R and AWK that has allowed us to speedup the querying from the original SPARK workflow by a large factor. It certainly wasn't easy and it took lots of iterations. This post is meant to help others avoid some of the same mistakes and show what did eventually work. 

First some background:

## The Data
The data was delivered to us by our university's genetics processing center as 25 TB of tsvs. Before handing it off to me my advisor split and gzipped these files into five batches each composed of roughly 240 four gigabyte files. Each row contained a data for a single SNP for a single person. `r tufte::margin_note("There were ~2.5 million SNPS and ~60 thousand people")` Along with the SNP value there were multiple numeric columns on things like intensity of the reading, frequency of different alleles etc. All told there were around 30 columns with frustratingly unique values. 

`r tufte::margin_note("<img src='https://www.researchgate.net/profile/Debi_Mishra4/publication/320298591/figure/fig5/AS:547880880820228@1507636389471/DNA-chips-to-analyze-single-nucleotide-polymorphisms-SNPs-A-DNA-chip-is-repaired-by.png' alt='DNA chips explainer' width=200px/> How a SNP Chip works. <a href='https://www.researchgate.net/figure/DNA-chips-to-analyze-single-nucleotide-polymorphisms-SNPs-A-DNA-chip-is-repaired-by_fig5_320298591'>Via Research Gate</a>")`



## The Goal
As with any data management project the most important thing is to consider _how_ the data will be used. In this case what __we will mostly be doing is fitting models and workflows on a SNP by SNP basis.__ I.e. we only will need a single SNP's data at a time. I needed to make it as easy, fast, and cheap as possible to extract all the records pertaining to one of the 2.5 million SNPs.

# How _not_ to do this
To appropriate a cliched quote: 

> I didn't fail a thousand times, I just discovered a thousand ways _not_ to parse lots of data into an easily queryable format. 

## The first attempt


[__Lesson Learned:__ There's no cheap way to parse 25tb of data at once.]{.lesson_learned}

Having taken a class at Vanderbilt titled 'Advanced methods in Big Data' I was sure I had this in the bag. `r tufte::margin_note("Capital B capital D Big Data, so you know it's serious.")`It would be maybe an hour or two of me setting up a Hive server to run over all our data and then calling it good.  Since our data is stored on AWS S3 I used a service called [Athena](https://aws.amazon.com/athena/) which allows you to run Hive SQL queries on your S3 data. Not only do you get to avoid setting/ spinning up a Spark cluster, you only pay for the data searched. 

After pointing Athena to my data and its format I ran a few tests with queries like 
```sql
select * from intensityData limit 10;
```

and got back results fast and well formed. I was set. 

Until we tried to use the data in real life....

I was asked to grab all the data for a SNP so we could test a model on it. I ran the query: 

```sql
select * from intensityData 
where snp = 'rs123456';
```


... and I waited. Eight minutes and 4+ terabytes of data queried later I had my results. Unfortunately, Athena charges you by data searched at the reasonable rate of \$5 per TB. So this single query cost \$20 and eight minutes.  `r tufte::margin_note("If we ever wanted to run a model over all the data we better be ready to wait roughly 38 years and pay $50 million.")` Clearly this wasn't going to work. 



## This should be a walk in the parquet...


[__Lesson Learned:__ Be careful with your parquet file sizes and organization.]{.lesson_learned}

My first attempt to remedy the situation was to convert all of the TSV's to [parquet files](https://parquet.apache.org). Parquet files are good for working with larger datasets because they store data in a 'columnar' fashion. Meaning each column is stored in its own section of memory/disk, rather than a text file with lines containing every column. This means to look for something you only have to read the neccesary column. Also,they keep a record of the range of values by column for each file so if the value you're looking for isn't in the column range Spark doesn't waste it's time scanning through the file. 

I ran a simple [AWS Glue job](https://aws.amazon.com/glue/) to convert our TSVs to Parquet and hooked up the new parquet files to Athena. Unfortunately, the query took just about the same amount of time and a tiny bit less money. This is because spark in its attempt to optimize the job just unzipped a single TSV chunk and placed it in its own parquet chunk. Because each chunk was big enough to contain multiple people's full records, this meant that every file had every SNP in them and thus Spark had to open all of them to extract what we wanted. 

![](/using_awk_and_r_to_parse_25tb/large_parquet_chunks.png)

## Sorting out the issue

[__Lesson Learned:__ Sorting is hard, especially when data is distributed.]{.lesson_learned}

I thought that I had the problem figured out now. All I needed to do was to sort the data on the SNP column instead of the individual. This would allow a given chunk of data to only have a few SNPs in it. Unfortunately, sorting billions of rows of data distributed across a cluster is not a trivial task.

<blockquote class="twitter-tweet" data-lang="en"><p lang="en" dir="ltr">Me taking algorithms class in college: &quot;Ugh, no one cares about computational complexity of all these sorting algorithms&quot; <br><br>Me trying to sort on a column in a 20TB <a href="https://twitter.com/hashtag/spark?src=hash&amp;ref_src=twsrc%5Etfw">#spark</a> table: &quot;Why is this taking so long?&quot; <a href="https://twitter.com/hashtag/DataScience?src=hash&amp;ref_src=twsrc%5Etfw">#DataScience</a> struggles.</p>&mdash; Nick Strayer (@NicholasStrayer) <a href="https://twitter.com/NicholasStrayer/status/1105127759318319105?ref_src=twsrc%5Etfw">March 11, 2019</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 


After attempting to run this on Amazon's glue it ran for 2 days and then crashed. 

## What about partitioning?

[__Lesson Learned:__ Partitions in spark need to be balanced.]{.lesson_learned} 

Another idea I had was to partition the data into Chromosomes. There are 23 of these (plus a few extra to account for mitochondrial DNA or unmapped regions etc). This would provide a way of cutting down the data into much more manageable chunks. By adding just a single line to the spark export function `partition_by = "chr"` the data should be put into those buckets. 

`r tufte::margin_note("<img src='https://cdn.prod-carehubs.net/n2/71b34990bba71dfd/uploads/2018/07/boldchromosomecellturquoise.png' alt='Chromosome graphic'/> DNA is made up of multiple chunks called Chromosomes. Img via [kintalk.org](https://kintalk.org/genetics-101/).")`

Unfortunately things didn't work out well. This is because the chromosomes are different sizes and thus have different amounts of data within them.  This meant that the tasks spark sent out to its workers were unbalanced and ran slowly due to some of the nodes finishing early and sitting idle. The jobs _did_ finish, however. Unfortunately for querying larger chromosomes (aka where we will actually want to get data) the cost was only improved ~10x. A lot but not enough.

## What about even finer partitioning? 

[__Lesson Learned:__ Never, ever, try and make 2.5 million partitions.]{.lesson_learned}

I decided to get crazy with my partitioning and partitioned on each SNP. This guaranteed that each partition would be equal in size. __THIS WAS A BAD IDEA.__ I used Glue and added the innocent line of `partition_by = 'snp'`. The job started and ran. A day later I checked and noticed nothing had been written to S3 yet so I killed the job. Turns out Glue _was_ writing intermediate files to hidden S3 locations, and a lot of them, like 2 billion. This mistake ended up costing more than a thousand dollars and didn't make my advisor happy. 

## Partitioning + Sorting

[__Lesson Learned:__ Sorting is still hard and so is tuning spark.]{.lesson_learned}

The last attempt in the partitioning era was to partition on chromosome and then sort each partition. This in theory would have made each query quicker because the desired SNP data would only reside in the ranges of a few of the parquet chunks within a given region. Unfortunately, it turns out sorting even the partitioned data was a lot of work. I ended up switching to EMR for a custom cluster, using 8 powerful instances (C5.4xl) and using Sparklyr to build a more flexible workflow...

```r
# sparklyr snippet to partition by chr and sort w/in partition
# Join the raw data with the snp bins
raw_data
  group_by(chr) %>%
  arrange(Position) %>% 
  spark_write_parquet(
    path = DUMP_LOC,
    mode = 'overwrite',
    partition_by = c('chr')
  )
```


...but no mater what Spark never finished. I tried all the tuning tricks: bumped up the memory allocated to each executor of the queries, used high ram node types, broadcasting variables, but it would always get around half way done then executors would slowly start failing till everything eventually ground to a halt. 

<blockquote class="twitter-tweet" data-lang="en"><p lang="en" dir="ltr">Current status: desperately watching <a href="https://twitter.com/hashtag/spark?src=hash&amp;ref_src=twsrc%5Etfw">#spark</a> UI to see stage progression. Will it complete half the steps and then drop all the executors with cryptic error messages and never spin them back up? Probably. Still better than game of thrones season 8. <a href="https://twitter.com/hashtag/DataScience?src=hash&amp;ref_src=twsrc%5Etfw">#DataScience</a> <a href="https://t.co/3Tc0a8PY2O">pic.twitter.com/3Tc0a8PY2O</a></p>&mdash; Nick Strayer (@NicholasStrayer) <a href="https://twitter.com/NicholasStrayer/status/1128702940376633344?ref_src=twsrc%5Etfw">May 15, 2019</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 


# Getting more creative

[__Lesson Learned:__ Sometimes bespoke bespoke data needs bespoke solutions.]{.lesson_learned}

Every SNP has a position value. This is an integer corresponding to how many bases along its chromosome it lies. This is a nice and natural method of organizing our data. The first though I had was building partitions by regions of each chromosome. Aka (positions 1 - 2000, 2001 - 4000, etc). The problem is SNPs are not evenly distributed along their chromosomes, so the bins would be wildly different in size. 


![](/using_awk_and_r_to_parse_25tb/uneven_snps.jpeg)

The solution I came up with was to bin by position _rank_. I ran a query on our already loaded data to get the list of the unique SNPs, their positions, and their chromosomes. I then sorted within each chromosome and bundled the SNPs into bins of a given size. E.g. 1000 SNPS. This gave me a mapping from SNP -> bin-in-chromosome. 

`r tufte::margin_note("I ended up using 75 SNPs per bin, I explain why later.")`

```r
snp_to_bin <- unique_snps %>% 
  group_by(chr) %>% 
  arrange(position) %>% 
  mutate(
    rank = 1:n()
    bin = floor(rank/snps_per_bin)
  ) %>% 
  ungroup()
```

## First attempt with Spark

[__Lesson Learned:__ Spark joining is fast, but partitioning is still expensive]{.lesson_learned}

The goal was to read this small (2.5 million row) dataframe into spark, join it with the raw data, and then partition on the newly added `bin` column. 

```r
# Join the raw data with the snp bins
data_w_bin <- raw_data %>%
  left_join(sdf_broadcast(snp_to_bin), by ='snp_name') %>%
  group_by(chr_bin) %>%
  arrange(Position) %>% 
  spark_write_parquet(
    path = DUMP_LOC,
    mode = 'overwrite',
    partition_by = c('chr_bin')
  )
```
`r tufte::margin_note("Notice the use of sdf_broadcast(), this lets spark know it should send this dataframe to all nodes. It's helpful when the data is small and needed for all tasks. Otherwise Spark tries to be clever and waits to distribute it till it needs it which can cause bottlenecks.")`


Unfortunately things didn't work out. Much like the sorting attempt the jobs would run for a while, finish the joining task, and then as the partitioning started executors would start crashing. 

## Bringing in AWK

[__Lesson Learned:__ Don't sleep on the basics. Someone probably solved your problem in the 80s.]{.lesson_learned}

Given that all my Spark failures were due to the data having to be shuffled around the different nodes as it was all mixed up, I figured I could attempt to help Spark out by preprocessing. I decided to try and split the raw text data on the chromosome column, that way I would be able to provide spark with some somewhat 'pre-partitioned' data. I stack overflow searched and found [this wonderful answer](https://unix.stackexchange.com/questions/114061/extract-data-from-a-file-and-place-in-different-files-based-on1-column-value) 

Using AWK you can split a text file up by values of a given column by actually performing the writing in the script rather than sending script result to `stdout`.

I wrote up a bash script to test this. I used the s3 commandline tool to download a batch of the gzipped files, then unzipped using `gzip`, piped that to `awk`. It worked! 

```{bash, eval = FALSE}
gzip -dc path/to/chunk/file.gz |
awk -F '\t' \
'{print $1",..."$30">"chunked/"$chr"_chr"$15".csv"}'
```


## Saturating the cores

[__Lesson Learned:__ `gnu parallel` is magic and everyone should use it.]{.lesson_learned}

The spliting was a tad bit slow and when I ran `htop` to see the usage of the powerful (expensive) ec2 instance I was using a single core and ~200 mb of ram. If I wanted to get things done and not waste a lot of money I was going to need to figure out how to parallize this. Luckily I found the chapter on parallelizing workflows in [Data Science at the Command Line](https://www.datascienceatthecommandline.com) and utterly fantastic book by Jeroen Janssens. It introduced me to `gnu parallel` which is very flexible method for spinning up multiple threads in a unix pipeline. 

`r tufte::margin_note("<a href = 'https://www.datascienceatthecommandline.com' ><img src = 'https://www.datascienceatthecommandline.com/images/cover.png' alt = 'Data Science at the Command Line book cover.'/></a>")`

Once I ran the splitting using the new GNU parallel workflow it was great, but I was still getting some bottlenecking caused by downloading the S3 objects being a little bit slow and not fully parallel. I did a few things to fix this. 

1. Found out that you can implement the s3 download step right into the pipeline, completely skipping intermediate disk storage (this was nice and meant I could use smaller and thus cheaper storage on AWS. 
2. Increased the number of threads that the AWS CLI uses to some large number (the default is 10) with `aws configure set default.s3.max_concurrent_requests 50`.
3. Switching to a network speed optimized ec2 instance. These are the ones with the `n` in the name. `r tufte::margin_note("I found that the loss in compute power caused from using the 'n' instances was more than made up for by the increased download speeds.")` I used c5n.4xl's for most of my stuff. 
4. Switched from `gzip` to `pigz`, which is a parallel gzip algorithm that does some clever things to parallelize an inherently unparallizable task of decompressing gziped files. (This helped the least)

```{bash, eval = FALSE}
# Let S3 use as many threads as it wants
aws configure set default.s3.max_concurrent_requests 50

for chunk_file in $(aws s3 ls $DATA_LOC | awk '{print $4}' | grep 'chr'$DESIRED_CHR'.csv') ; do

        aws s3 cp s3://$batch_loc$chunk_file - |
        pigz -dc |
        parallel --block 100M --pipe  \
        "awk -F '\t' '{print \$1\",...\"$30\">\"chunked/{#}_chr\"\$15\".csv\"}'"

       # Combine all the parallel process chunks to single files
        ls chunked/ |
        cut -d '_' -f 2 |
        sort -u |
        parallel 'cat chunked/*_{} | sort -k5 -n -S 80% -t, | aws s3 cp - '$s3_dest'/batch_'$batch_num'_{}'
		
		 # Clean up intermediate data
       rm chunked/*
done
```

These steps combined to make things _very_ fast. By virtue of increasing the speed of download and avoiding writing to disk I was now able to process a whole 5 terabyte batch in just a few hours. 

<blockquote class="twitter-tweet" data-lang="en"><p lang="en" dir="ltr">There&#39;s nothing sweeter than seeing all the cores you&#39;re paying for on AWS being used. Thanks to gnu-parallel I can unzip and split a 19gig csv just as fast as I can download it. I couldn&#39;t even get spark to run this. <a href="https://twitter.com/hashtag/DataScience?src=hash&amp;ref_src=twsrc%5Etfw">#DataScience</a> <a href="https://twitter.com/hashtag/Linux?src=hash&amp;ref_src=twsrc%5Etfw">#Linux</a> <a href="https://t.co/Nqyba2zqEk">pic.twitter.com/Nqyba2zqEk</a></p>&mdash; Nick Strayer (@NicholasStrayer) <a href="https://twitter.com/NicholasStrayer/status/1129416944233226240?ref_src=twsrc%5Etfw">May 17, 2019</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 

`r tufte::margin_note("This tweet should have said 'tsv'. Alas.")`



## Using newly parsed data

[__Lesson Learned:__ Spark likes uncompressed data and does not like combining partitions.]{.lesson_learned}

Now that I had the data sitting in an unzipped (see splittable) and semi-organized format on S3 I could go back to spark and work again. Surprise -- things didn't workout again! It was very hard to accurately tell spark how the data was partitioned and even when I did it seemed to like to split things into way too many partitions (like 95k), which then when I used `coalesce` to reduce down to a reasonable number of partitions, ended up ruining the partitioning I had used. I am sure there is a way to fix this but I couldn't find it over a couple days of looking. I did end up getting things to finish on Spark, it took a while however and my split parquet files were not super tiny (~200kb) But the data was where it needed to be. 

<img src='/using_awk_and_r_to_parse_25tb/uneven_small_partitions.png'/> 
[_Too small and uneven, wonderful!_]{.photo_caption}


## Testing out local spark queries

[__Lesson Learned:__ Spark is a lot of overhead for simple jobs.]{.lesson_learned}

Now that the data was loaded up into a reasonable format I could test out the  speed. I setup an R script to spin up a local spark server and then load a spark dataframe from a given parquet bins location. `r tufte::margin_note("I tried loading all the data but couldn't get sparklyr to recognize the partitioning for some reason.")`


```r
sc <- spark_connect(master = "local")

desired_snp <- 'rs34771739'

# Start a timer
start_time <- Sys.time()

# Load the desired bin into spark
intensity_data <- sc %>% 
  spark_read_parquet(
    name = 'intensity_data', 
    path = get_snp_location(desired_snp),
    memory = FALSE )

# Subset bin to snp and then collect to local
test_subset <- intensity_data %>% 
  filter(SNP_Name == desired_snp) %>% 
  collect()

print(Sys.time() - start_time)
``` 

This took 29.415 seconds. Much better than before, but still not a great sign for mass testing of anything. In addition, I couldn't try and speed it up by enabling caching because when I tried to cache the bin's Spark dataframe in memory Spark always crashed, even when I gave it 50+ gigs of memory for a dataset that was at this point smaller than 15. 

## Back to AWK

[__Lesson Learned:__ Associative arrays in AWK are super powerful.]{.lesson_learned}

I knew I could do better. I remembered that I had read about a cool feature in AWK called ["associative arrays'](http://www.grymoire.com/Unix/Awk.html#uh-22). in this charming [AWK guide by Bruce Barnett](http://www.grymoire.com/Unix/Awk.html). These are essentially a key value stores in AWK that for some reason had been given a different name and thus I never thought too much about. I realized that I could use these associative arrays to perform the union between my SNP -> bin table and my raw data without using something like spark. 

What I had do to was use the `BEGIN` block in my AWK script. This is a block of code that gets run before any lines of data are fed into the main body of the script. I used this block to read my entire SNP -> bin table in to an associative array that was keyed with the SNP name and valued with the SNPs bin. 

__join_data.awk__
```awk
BEGIN {
  FS=",";
  batch_num=substr(chunk,7,1);
  chunk_id=substr(chunk,15,2);
  while(getline < "snp_to_bin.csv") {bin[$1] = $2}
}
{
  print $0 > "chunked/chr_"chr"_bin_"bin[$1]"_"batch_num"_"chunk_id".csv"
}
```


The `while(getline...)` command loaded all the rows in from my bin csv and set the first column (the SNP name) as the key to the `bin` associative array and the second value (the bin) to the value. Then in the `{` block `}` that gets run on every line of the main file each line is sent to an output file that was had a unique name based upon its bin: `..._bin_"bin[$1]"_...`. 

`r tufte::margin_note("The variables of <code>batch_num</code> and <code>chunk_id</code> corresponded to data given by the pipeline that allowed me to avoid race conditions in my writing by making sure that every thread run by <code>parallel</code> wrote to its own unique file. ")`


Because I had all the raw data split into chromosome folders from my previous AWK experiment I could now write another bash script to work through a chromosome at a time and send back the now further partitioned data into S3. 

```{bash, eval = FALSE}
DESIRED_CHR='13'

# Download chromosome data from s3 and split into bins
aws s3 ls $DATA_LOC |
awk '{print $4}' |
grep 'chr'$DESIRED_CHR'.csv' |
parallel "echo 'reading {}'; aws s3 cp "$DATA_LOC"{} - | awk -v chr=\""$DESIRED_CHR"\" -v chunk=\"{}\" -f split_on_chr_bin.awk"

# Combine all the parallel process chunks to single files and upload to rds using R
ls chunked/ |
cut -d '_' -f 4 |
sort -u |
parallel "echo 'zipping bin {}'; cat chunked/*_bin_{}_*.csv | ./upload_as_rds.R '$S3_DEST'/chr_'$DESIRED_CHR'_bin_{}.rds"
rm chunked/*
```

This script has two `parallel` sections:

The first one splits up every file containing the desired chromosome's data and spins up a separate process to spit that file into its representative bins. In order to prevent race conditions from the multiple threads writing to the same bin file AWK is passed the name of the file which it uses to write to unique locations. This results in a _ton_ of tiny files located on the disk (I used 1tb EBS volumes for this). 

The second `parallel` pipeline goes through and merges every bin's separate files into single csv's with `cat`, sends them to an R script... 

## Piping to R?

[__Lesson Learned:__ You can access `stdin` and `stdout` from inside an R script.]{.lesson_learned}

You may have noticed this part of the bash script above: `...cat chunked/*_bin_{}_*.csv | ./upload_as_rds.R...`. This line pipes all the concated files for a bin into the following R script...

```r
#!/usr/bin/env Rscript
library(readr)
library(aws.s3)

# Read first command line argument
data_destination <- commandArgs(trailingOnly = TRUE)[1]

data_cols <- list(SNP_Name = 'c', ...)

s3saveRDS(
  read_csv(
		file("stdin"), 
  		col_names = names(data_cols),
  		col_types = data_cols 
	),
  object = data_destination
)
```

By passing `readr::read_csv` the value `file("stdin")` it loads the data piped to the R script into a dataframe, which then gets written as an `.rds` file directly to s3 using `aws.s3`. 

`r tufte::margin_note("Rds is kind-of like a junior version of parquet without the niceties of columnar storage.")`

After this bash script had finishes I have a bunch of `.rds` files sitting in S3 benefiting from the great compression techniques and also types built in that `.rds` offers.  

Even with notoriously slow R in the workflow this was super fast. `r tufte::margin_note("Unsurprisingly the parts of R for reading and writing data are rather optimized.")` After testing on a single average size chromosome the job finished in about two hours. 


## Limits of S3

[__Lesson Learned:__ S3 can handle a _lot_ of files due to smart path implementation.]{.lesson_learned}

I was worried about how S3 would handle having a ton of files dumped onto it. I could make the file names make sense, but how would S3 handle searching for one? 

It turns out S3 treats the path to a given file as a simple key in what can be thought of as a hash table. Aka a document-based database. Think of a "bucket" as a table and every file is an entry. Because of this folders in S3 are just a cosmetic thing and S3 doesn't actually care about the `/` character. 

`r tufte::margin_note("<img src='/using_awk_and_r_to_parse_25tb/s3_keys_explained.png'/> <a href='https://aws.amazon.com/s3/faqs/'>From the S3 FAQs page</a>")`

Because speed and efficiency are a important to S3 making money for Amazon it's no surprise that this key-is-a-file-path system is super optimized. Still I tried to strike a balance. I wanted to not need to do a ton of `get` requests and I wanted the queries to be fast. I found that making around 20k bin files worked best. `r tufte::margin_note("I am sure further optimizations could speed things up (such as making a special bucket just for the data and thus reducing the size of the lookup table.) But I ran out of time and money to do more experiments.")`

## What about cross-compatability?

[__Lesson Learned:__ Premature optimizing your storage method is the root of all time wasted.]{.lesson_learned}

A very reasonable thing to ask at this point is "why would you use a proprietary file format for this?" The reason came down to speed of loading (using gzipped csvs took about 7 times longer to load) and compatibility with our workflows.`r tufte::margin_note("Once R can easily load Parquet (or Arrow) files without the overhead of spark I may reconsider.")` Everyone else in my lab exclusively uses R and if I end up needing to convert the data to another format I still have the original raw text data and can just run the pipeline again. 

## Divying out the work

[__Lesson Learned:__ Don't try to hand optimize jobs, let the computer do it.]{.lesson_learned}

I wanted to spin up multiple ec2 instances to convert all my data but I also didn’t want to have super unbalanced job loads (just like how spark suffered from the unbalanced partitions). `r tufte::margin_note("I also didn’t want to spin up a single instance for each chromosome, since there is a limit by default of 10 instances at a time for AWS accounts.")` 

My solution was to write a brute force job optimization script using R….

First I queried S3 to figure out how large each chromosome was in terms of storage.
```r
library(aws.s3)
library(tidyverse)

chr_sizes <- get_bucket_df(
  bucket = '...', prefix = '...', max = Inf
) %>% 
  mutate(Size = as.numeric(Size)) %>% 
  filter(Size != 0) %>% 
  mutate(
    # Extract from the file name the chromosome
    chr = str_extract(Key, 'chr.{1,4}\\.csv') %>%
			 str_remove_all('chr|\\.csv')
  ) %>% 
  group_by(chr) %>% 
  summarise(total_size = sum(Size)/1e+9)
```

```
# A tibble: 27 x 2
   chr   total_size
   <chr>      <dbl>
 1 0           163.
 2 1           967.
 3 10          541.
 4 11          611.
 5 12          542.
 6 13          364.
 7 14          375.
 8 15          372.
 9 16          434.
10 17          443.
# … with 17 more rows
```

Then I wrote a function that would take this total size info, shuffle the order, and split into `num_jobs` groups and report how variable the sizes of each job’s data was. 

```r
num_jobs <- 7
# How big would each job be if perfectly split?
job_size <- sum(chr_sizes$total_size)/7

shuffle_job <- function(i){
  chr_sizes %>%
    sample_frac() %>% 
    mutate(
      cum_size = cumsum(total_size),
      job_num = ceiling(cum_size/job_size)
    ) %>% 
    group_by(job_num) %>% 
    summarise(
      job_chrs = paste(chr, collapse = ','),
      total_job_size = sum(total_size)
    ) %>% 
    mutate(sd = sd(total_job_size)) %>% 
    nest(-sd)
}

shuffle_job(1)
```
```
# A tibble: 1 x 2
     sd data            
  <dbl> <list>          
1  153. <tibble [7 × 3]>
```


Once this was setup all I had do do was run a ton of shuffles and pick the best one using purrr. 
```r
1:1000 %>% 
  map_df(shuffle_job) %>% 
  filter(sd == min(sd)) %>% 
  pull(data) %>% 
  pluck(1)
```

This gave me a set of jobs that were all very close in size. All I had do do then was wrap my previous bash script in a big for loop…
`r tufte::margin_note("It took me ~10 mins to write this job optimization which was way less time than the inballance caused by my manual job creation would have added to the processing so I think I didn't fall for premature optimization here.")`

```{bash, eval = FALSE}
for DESIRED_CHR in "16" "9" "7" "21" "MT"
do
#...
fi
```

add a shutdown command at the end….
```{bash, eval = FALSE}
sudo shutdown -h now
```

… and I was off the the races. I now used the AWS CLI to spin up a bunch of instances, passing them this bash script via the `user_data` option. They ran and then shutdown automatically so I didn’t pay for extra compute. 

```{bash, eval = FALSE}
aws ec2 run-instances ...\
--tag-specifications "ResourceType=instance,Tags=[{Key=Name,Value=<<job_name>>}]" \
--user-data file://<<job_script_loc>>
```
 
# Packing it up!
[__Lesson Learned:__ Keep API simple for your end users and flexible for you.]{.lesson_learned}

Finally, the data was where and how I needed it. The last step was to simplify the process for using the data as much as possible for my lab members. I wanted to provide a simple API for querying. If in the future I did decide to switch from using `.rds` to parquet files I wanted to be able to make that my issue and not my lab mate's. The way I decided to to this was an internal R package. 

I built and documented a very simple package that contains just a few functions for accessing the data, centered around the function `get_snp`. In addition I built a [pkgdown site](https://pkgdown.r-lib.org) so lab members could easily see examples/docs. 


<img src='/using_awk_and_r_to_parse_25tb/pkg_down.png'/> 

## Intelligent caching. 

[__Lesson Learned:__ If your data is setup well caching will be easy!]{.lesson_learned}

Since one of the main workflows for these data was running the same model/ analysis across a bunch of SNPs at a time I decided that I could use the binning to my advantage. When pulling the data down for a SNP the entire bin's data is kept and attached to the returned object. This means if a new query is run the old queries result's can (potentially) be used to speed it up.

```r
# Part of get_snp()
...
  # Test if our current snp data has the desired snp.
  already_have_snp <- desired_snp %in% prev_snp_results$snps_in_bin

  if(!already_have_snp){
    # Grab info on the bin of the desired snp
    snp_results <- get_snp_bin(desired_snp)

    # Download the snp's bin data
    snp_results$bin_data <- aws.s3::s3readRDS(object = snp_results$data_loc)
  } else {
    # The previous snp data contained the right bin so just use it
    snp_results <- prev_snp_results
  }
...
```

Since a common workflow is scanning SNP-by-SNP along a region, caching the bin made it very likely that the next requested SNP would already be on the machine and thus could be immediately grabbed. 


`r tufte::margin_note("While building the package I ran a lot of benchmarks to compare the speed between different methods. I recomend it because sometimes the results went against my intuition. For instance, <code>dplyr::filter</code> was much faster than using indexing based filtering for grabbing rows, but getting a single column from a filtered dataframe was much faster using indexing syntax.")`

Notice that the `prev_snp_results` object contains the key `snps_in_bin`. This allows fast searching through if we already have the data from a previous query, while also making it easy for the user to loop through all the SNPs in a bin using code like: 

```r
# Get bin-mates
snps_in_bin <- my_snp_results$snps_in_bin

for(current_snp in snps_in_bin){
  my_snp_results <- get_snp(current_snp, my_snp_results)
  # Do something with results 
}
```

# End results

We are now able to (and have started in ernest) run models and scenarios we were incapable of before. The best part is the other members of my lab don’t have to think about the complexities that went into it. They just have a function that works. 

The speed is much improved. A typical use-case is to scan a functionally significant region of the genome (such as a gene) . Before we couldn’t do this (because it cost too much) but now, because of the caching, it takes on average less than a tenth of a second per SNP queried and the data usage is not even high enough to round up to a penny on our S3 costs. 

<blockquote class="twitter-tweet" data-lang="en"><p lang="en" dir="ltr">Recently I got put in change of wrangling 25+ TB of raw genotyping data for my lab. When I started, using spark took 8 min &amp; cost $20 to query a SNP. After using AWK + <a href="https://twitter.com/hashtag/rstats?src=hash&amp;ref_src=twsrc%5Etfw">#rstats</a> to process, it now takes less than a 10th of a second and costs $0.00001. My personal <a href="https://twitter.com/hashtag/BigData?src=hash&amp;ref_src=twsrc%5Etfw">#BigData</a> win. <a href="https://t.co/ANOXVGrmkk">pic.twitter.com/ANOXVGrmkk</a></p>&mdash; Nick Strayer (@NicholasStrayer) <a href="https://twitter.com/NicholasStrayer/status/1134151057385369600?ref_src=twsrc%5Etfw">May 30, 2019</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 


## Take Away


This post post isn't meant to be a how-to guide. The final solution is bespoke and almost assuredly not the optimal one. For risk of sounding unbearably cheesy this was about the journey. I want others to realize that these solutions don't pop fully formed into peoples head's but they are a product of trial and error. 

In addition, if you are in the position of hiring someone as a data scientist please consider the fact that getting good at these tools requires experience, and experience requires money. I am lucky that I have grant funding to pay for this but many who assuredly could do a better job than me will never get the chance because they don't have the funds to even try. 

"Big Data" tools are generalists. If you have the time you will almost assuredly be able to write up a faster sollution to your problem using smart data cleaning, storage, and retrieval techniques. Ultimately it comes down to a cost-benefit analysis.


